docs/contributing/retries-and-waiters.md

```diff
@@ -0,0 +1,471 @@
+
+```go
+// aws/internal/service/example/waiter/status.go (created if does not exist)
+
+// ThingAttribute fetches the Thing and its Attribute
+func ThingAttribute(conn *example.Example, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		output, err := /* ... AWS Go SDK operation to fetch resource/value ... */
+
+		if tfawserr.ErrCodeEquals(err, example.ErrCodeResourceNotFoundException) {
+			return nil, "", nil
+		}
+
+		if err != nil {
+			return nil, "", err
+		}
+
+		if output == nil {
+			return nil, "", nil
+		}
+
+		return output, aws.StringValue(output.Attribute), nil
+	}
+}
+```
+
+```go
+// aws/internal/service/example/waiter/waiter.go (created if does not exist)
+
+const (
+	ThingAttributePropagationTimeout = 2 * time.Minute
+)
+
+// ThingAttributeUpdated is an attribute waiter for ThingAttribute
+func ThingAttributeUpdated(conn *example.Example, id string, expectedValue string) (*example.Thing, error) {
+	stateConf := &resource.StateChangeConf{
+		Target:  []string{expectedValue},
+		Refresh: ThingAttribute(conn, id),
+		Timeout: ThingAttributePropagationTimeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if output, ok := outputRaw.(*example.Thing); ok {
+		return output, err
+	}
+
+	return nil, err
+}
+```
+
+```go
+// aws/resource_example_thing.go
+
+function ExampleThingUpdate(d *schema.ResourceData, meta interface{}) error {
+	// ...
+
+	d.HasChange("attribute") {
+		// ... AWS Go SDK logic to update attribute ...
+
+		if err := waiter.ThingAttributeUpdated(conn, d.Id(), d.Get("attribute").(string)); err != nil {
+			return fmt.Errorf("error waiting for Example Thing (%s) attribute update: %w", d.Id(), err)
+		}
+	}
+
+	// ...
+}
+```
+
+## Asynchronous Operations
+
+Many longer running operations are initiated and performed asynchronously in the background after returning a successful response to the initial operation request. These operations may either have a component level status field that can be tracked (e.g. `CREATING`, `UPDATING`, etc.) or they may have an explicit tracking identifier to follow the progress of the background operation.
+
+Terraform resources are expected to wait for these background operations to complete. Failing to do so may introduce incomplete state information or downstream errors in other resources. In rare scenarios involving very long operations, operators may request a flag to skip the waiting, however these should only be implemented case-by-case to prevent those previously mentioned confusing issues.
+
+### AWS Go SDK Waiters
+
+In limited cases, the AWS service API model includes the information to automatically generate a waiter function in the AWS Go SDK for an operation. These are typically named with the prefix `WaitUntil...`. If available, these functions are often good enough for an initial resource implementation and can be used. For example:
```

Comment:
The term "model" here is important. These are not part of the API itself. Also I think distinctively saying limited (as they are rare) versus some is also important since they are hard to find.

```suggestion
In limited cases, the AWS service API model includes the information to automatically generate a waiter function in the AWS Go SDK for an operation. These are typically named with the prefix `WaitUntil...`. If available, these functions can be used for an initial resource implementation. For example:
```